### ISSUE TITLE
[Perf regression] pprof flame graphs reveal Link is now very expensive.

### ISSUE THREAD

**Issue by selasijean** [wcharczuk/go-incr#7]
Calculations for 600 entities that used to take < 500 milliseconds now take 10+ minutes. Did some profiling and found some interesting bottlenecks. Main culprit - DetectCycleIfLinked

<img width="1728" alt="Screenshot 2024-02-06 at 2 08 52 AM" src="https://github.com/wcharczuk/go-incr/assets/19496572/929afad1-762f-463b-a6f4-a1042ec017a3">

<img width="1728" alt="Screenshot 2024-02-06 at 2 11 56 AM" src="https://github.com/wcharczuk/go-incr/assets/19496572/39eb6d49-b9e7-4292-940d-b8ae1fac7c6b">

You'll find the pprof file [here](https://drive.google.com/file/d/1ZzIepcxKnftf1kJBFByhISwFIXIf3-sq/view?usp=sharing) if you want to dig in more

```
 go tool pprof -http=: [filename]
```

**Comment by wcharczuk** [comment]
I've tweaked `Link` to not have a dedicated `DetectCycleIfLinked` call, and instead just checks for duplicate IDs in the propagate height change pass. Let me know if that helps, otherwise can dig in more.

**Comment by selasijean** [comment]
~75% improvement. ~12mins -> ~3mins .. 

<img width="1728" alt="Screenshot 2024-02-06 at 7 54 38 AM" src="https://github.com/wcharczuk/go-incr/assets/19496572/bda5fac2-f088-439f-8ae7-3e35e15ac8c9">

**Comment by wcharczuk** [comment]
It's still spending a lot of time in `Link`, specifically in calculating heights. 

I used to calculate height on observation using a full depth pass, but switched to doing it on link to make the depth calculation simpler, it may bear reverting that. I'll look more later today.

**Comment by selasijean** [comment]
I'm actually surprised how expensive maps are in go. Wow. Maybe we can do without it and use slices? .. It seems maps are useful only for Remove, RemoveKey, HasKey, Has ..
Maybe we manage both map and list - using the map for the lookups. Remove will be painful for the slice though

**Comment by selasijean** [comment]
Actually maybe a linkedlist instead

**Comment by wcharczuk** [comment]
Took a stab at refactoring the node lists as arrays, if you want to give it a shot it's on a branch `will--node-slices`

**Comment by selasijean** [comment]
It got much worse. The simple formula, f(t) = f(t-1) (bind(t) -> bind(t-1)) is exhibiting an interesting behavior in this case. 

Time to compute(t )= ~2 * Time to compute(t-1) .. It's basically getting very slow with time. We seem to be doing twice as much work somewhere in the code.


<img width="1685" alt="Screenshot 2024-02-06 at 11 10 46 PM" src="https://github.com/wcharczuk/go-incr/assets/19496572/2ded6ecb-e6b0-4ee7-893a-e647b02891b5">

**Comment by wcharczuk** [comment]
```$ go test -bench=.                                                                                                                                          go-incr main
goos: darwin
goarch: arm64
pkg: github.com/wcharczuk/go-incr
Benchmark_Stabilize_withPreInitialize_512-12                        3544            334136 ns/op
Benchmark_Stabilize_withPreInitialize_1024-12                       1701            686518 ns/op
Benchmark_Stabilize_withPreInitialize_2048-12                        849           1409207 ns/op
Benchmark_Stabilize_withPreInitialize_4096-12                        416           2821148 ns/op
Benchmark_Stabilize_withPreInitialize_8192-12                        160           6659809 ns/op
Benchmark_Stabilize_withPreInitialize_16384-12                        58          17498667 ns/op
Benchmark_ParallelStabilize_withPreInitialize_512-12                5242            206747 ns/op
Benchmark_ParallelStabilize_withPreInitialize_1024-12               2972            413081 ns/op
Benchmark_ParallelStabilize_withPreInitialize_2048-12               1344            878369 ns/op
Benchmark_ParallelStabilize_withPreInitialize_4096-12                717           1728638 ns/op
Benchmark_ParallelStabilize_withPreInitialize_8192-12                364           3371231 ns/op
Benchmark_ParallelStabilize_withPreInitialize_16384-12               154           7700844 ns/op
Benchmark_Stabilize_deep_2_32-12                                  130470              8657 ns/op
Benchmark_Stabilize_deep_4_64-12                                   61016             19832 ns/op
Benchmark_Stabilize_deep_4_96-12                                   37338             32200 ns/op
Benchmark_Stabilize_deep_4_128-12                                  25688             45757 ns/op
Benchmark_Stabilize_deep_8_256-12                                   9468            105664 ns/op
Benchmark_Stabilize_deep_8_512-12                                   4216            243333 ns/op
Benchmark_Stabilize_nestedBinds_64-12                               3104            364077 ns/op
Benchmark_Stabilize_nestedBinds_128-12                              1585            762989 ns/op
Benchmark_Stabilize_nestedBinds_256-12                               633           1920651 ns/op
Benchmark_Stabilize_nestedBinds_512-12                               237           5003219 ns/op
Benchmark_Stabilize_nestedBinds_1024-12                               82          14884032 ns/op
PASS
ok      github.com/wcharczuk/go-incr    33.639s```

this is on `main`

now in the branch:

```$ go test -bench=.                                                                                                                            go-incr will--node-slices U
goos: darwin
goarch: arm64
pkg: github.com/wcharczuk/go-incr
Benchmark_Stabilize_withPreInitialize_512-12                        5047            235604 ns/op
Benchmark_Stabilize_withPreInitialize_1024-12                       2192            521280 ns/op
Benchmark_Stabilize_withPreInitialize_2048-12                       1087           1070649 ns/op
Benchmark_Stabilize_withPreInitialize_4096-12                        553           2137411 ns/op
Benchmark_Stabilize_withPreInitialize_8192-12                        232           4858307 ns/op
Benchmark_Stabilize_withPreInitialize_16384-12                       102          10985270 ns/op
Benchmark_ParallelStabilize_withPreInitialize_512-12                6321            194922 ns/op
Benchmark_ParallelStabilize_withPreInitialize_1024-12               3174            379615 ns/op
Benchmark_ParallelStabilize_withPreInitialize_2048-12               1224            826760 ns/op
Benchmark_ParallelStabilize_withPreInitialize_4096-12                802           1610700 ns/op
Benchmark_ParallelStabilize_withPreInitialize_8192-12                400           3034312 ns/op
Benchmark_ParallelStabilize_withPreInitialize_16384-12               186           6039694 ns/op
Benchmark_Stabilize_deep_2_32-12                                  165904              6728 ns/op
Benchmark_Stabilize_deep_4_64-12                                   84187             14356 ns/op
Benchmark_Stabilize_deep_4_96-12                                   52000             23269 ns/op
Benchmark_Stabilize_deep_4_128-12                                  34252             33929 ns/op
Benchmark_Stabilize_deep_8_256-12                                  12408             94424 ns/op
Benchmark_Stabilize_deep_8_512-12                                   5383            208085 ns/op
Benchmark_Stabilize_nestedBinds_64-12                            8714553               135.3 ns/op
Benchmark_Stabilize_nestedBinds_128-12                           8460531               135.3 ns/op
Benchmark_Stabilize_nestedBinds_256-12                           8174994               138.7 ns/op
Benchmark_Stabilize_nestedBinds_512-12                           7199878               145.5 ns/op
Benchmark_Stabilize_nestedBinds_1024-12                          6166285               163.8 ns/op
PASS
ok      github.com/wcharczuk/go-incr    34.881s
```

**Comment by wcharczuk** [comment]
there is a fair bit of work now for unobserves; specifically we have to plumb the graph for _anywhere_ that still references a node to prevent a specific breakage i've added some regression tests for (see the other issue), i'll get creative with how to approach this is the perf is still bad

**Comment by wcharczuk** [comment]
#9 to address

**Comment by selasijean** [comment]
Dropped to ~2:30min

**Comment by selasijean** [comment]
Example of what I meant

```
func Benchmark_Stabilize_connectedGraph_with_nestedBinds_1024(b *testing.B) {
	benchmarkConnectedGraphWithNestedBinds(1024, b)
}

func benchmarkConnectedGraphWithNestedBinds(depth int, b *testing.B) {
	ctx := context.Background()
	fakeFormula := Var(Root(), "fakeFormula")
	observed := make([]ObserveIncr[*int], depth)
	g := New(
		GraphMaxRecomputeHeapHeight(1024),
	)
	for x := 0; x < b.N; x++ {
		for i := 0; i < depth; i++ {
			o := makeSimpleNestedBindGraph(g, depth, fakeFormula)
			observed[i] = o
		}
		err := g.Stabilize(ctx)
		if err != nil {
			b.Error(err)
			b.FailNow()
		}
		for _, o := range observed {
			if o.Value() == nil {
				b.FailNow()
			}
		}
	}
}

func makeSimpleNestedBindGraph(graph *Graph, depth int, fakeFormula VarIncr[string]) ObserveIncr[*int] {
	cache := make(map[string]Incr[*int])
	var f func(bs *BindScope, t int) Incr[*int]
	f = func(bs *BindScope, t int) Incr[*int] {
		key := fmt.Sprintf("f-%d", t)
		if _, ok := cache[key]; ok {
			return WithinBindScope(bs, cache[key])
		}
		r := Bind(Root(), fakeFormula, func(bs *BindScope, formula string) Incr[*int] {
			if t <= 0 {
				out := 0
				r := Return(bs, &out)
				r.Node().SetLabel("f-0")
				return r
			}
			return Map(bs, f(bs, t-1), func(r *int) *int {
				out := *r
				return &out
			})
		})

		r.Node().SetLabel(fmt.Sprintf("f(%d)", t))
		cache[key] = r
		return r
	}
	o := f(Root(), depth)
	return Observe(Root(), graph, o)
}

```

**Comment by wcharczuk** [comment]
Thanks for the test case!

So this is still super expensive, but I've tuned things a bit to get that specific test to be a little faster.

I think there is still a lot of work to do, and separately I'm looking into how to integrate an adjustHeightsHeap.

Lemme know how this performs on latest main.

**Comment by wcharczuk** [comment]
OK the refactor is in, I'll be tuning things and adding tests over the weekend, but so far it looks a little better.

One quick note about the above test; it creates a _lot_ of nodes, specifically because it mutates a graph over and over within the benchmark N loop, even refactoring it not to do that, it still creates a quadratic number of nodes which becomes kind of bad past 256^3 (e.g. 16 million nodes). 

there is some improvement though, and hopefully you see times come down a bit more.

**Comment by selasijean** [comment]
Times have indeed significantly come down. We back to milliseconds 🙌  but in some cases however it spikes back up to minutes. I think these test cases repro the issue. It seems the order in which you do the calculations can significantly affect perf.

**Formulas**

```
	f = func(bs *incr.BindScope, t int) incr.Incr[*int] {
		key := fmt.Sprintf("f-%d", t)
		if _, ok := cache[key]; ok {
			return incr.WithinBindScope(bs, cache[key])
		}

		r := incr.Bind(incr.Root(), fakeFormula, func(bs *incr.BindScope, formula string) incr.Incr[*int] {
			if t <= 0 {
				out := 0
				r := incr.Return(bs, &out)
				r.Node().SetLabel("f-0")
				return r
			}
			bindOutput := incr.Map(bs, f(bs, t-1), func(r *int) *int {
				if r == nil {
					return nil
				}
				out := *r + 1
				return &out
			})
			bindOutput.Node().SetLabel(fmt.Sprintf("map-f-%d", t))
			return bindOutput
		})

		r.Node().SetLabel(fmt.Sprintf("f(%d)", t))
		cache[key] = r
		return r
	}

	// burn(t) = f(t)
	burn := func(bs *incr.BindScope, t int) incr.Incr[*int] {
		return incr.Bind(bs, fakeFormula, func(bs *incr.BindScope, formula string) incr.Incr[*int] {
			return f(bs, t)
		})
	}

	// cashbalance = cashbalance(t-1) - burn(t)
	var cash_balance func(bs *incr.BindScope, t int) incr.Incr[*int]
	cash_balance = func(bs *incr.BindScope, t int) incr.Incr[*int] {
		return incr.Bind(bs, fakeFormula, func(bs *incr.BindScope, formula string) incr.Incr[*int] {
			if t <= 0 {
				out := 0
				r := incr.Return(bs, &out)
				return r
			}
			return incr.Map2(bs, cash_balance(bs, t-1), burn(bs, t), func(c *int, b *int) *int {
				if c == nil || b == nil {
					return nil
				}
				out := *c - *b
				return &out
			})
		})
	}

	// monthofrunway = if burn > 0 then cashbalance / burn else 0
	months_of_runway := func(bs *incr.BindScope, t int) incr.Incr[*int] {
		return incr.Bind(bs, fakeFormula, func(bs *incr.BindScope, formula string) incr.Incr[*int] {
			zero := 0
			predicateIncr := incr.Map2(bs, burn(bs, t), incr.Return(bs, &zero), func(val *int, cmp *int) bool {
				if val == nil || cmp == nil {
					return false
				}
				return *val > *cmp
			})
			return incr.Bind(bs, predicateIncr, func(bs *incr.BindScope, predicate bool) incr.Incr[*int] {
				var out int = 0
				if predicate {
					return incr.Map2(bs, cash_balance(bs, t), burn(bs, t), func(c *int, b *int) *int {
						if c == nil || b == nil {
							return nil
						}
						out = *c / *b
						return &out
					})
				}
				return incr.Return(bs, &out)
			})
		})
	}
```

### Tests

**Test 1:** Calculate months_of_runway then burn.
```
t.Run("month_of_runway = if burn > 0: cash_balance / burn else 0. Calculate months of runway then burn", func(t *testing.T) {
		graph := incr.New()

		num := 24

		graph = incr.New()

		fmt.Println("Calculating months of runway for t= 1 to 24")
		start := time.Now()
		for i := 0; i < num; i++ {
			o := months_of_runway(incr.Root(), i)
			incr.Observe(incr.Root(), graph, o)
		}

		graph.Stabilize(ctx)
		elapsed := time.Since(start)
		fmt.Printf(fmt.Sprintf("Calculating months of runway took %s \n", elapsed))

		fmt.Println("Calculating burn for t= 1 to 24")
		start = time.Now()
		for i := 0; i < num; i++ {
			o := burn(incr.Root(), i)
			incr.Observe(incr.Root(), graph, o)
		}

		graph.Stabilize(ctx)
		elapsed = time.Since(start)
		fmt.Printf(fmt.Sprintf("Calculating burn took %s \n", elapsed))

		assert.Fail(t, "deliberate")
	})

Calculating months of runway for t= 1 to 24
Calculating months of runway took 18.769292ms 
Calculating burn for t= 1 to 24
Calculating burn took 806.290833ms 
```
Calculating months of runway involves calculating burn implicitly. Then in the next iteration, burn is calculated and we'd expect that it is quicker because calculating months of runway should have cached some burn results. 

**Test 2**: Calculate burn then months_of_runway
```
	t.Run("month_of_runway = if burn > 0: cash_balance / burn else 0. Calculate burn then months of runway", func(t *testing.T) {
		graph := incr.New()

		num := 24 // 200 out of adjust head error

		graph = incr.New()

		fmt.Println("Calculating burn for t= 1 to 24")
		start := time.Now()
		for i := 0; i < num; i++ {
			o := burn(incr.Root(), i)
			incr.Observe(incr.Root(), graph, o)
		}

		graph.Stabilize(ctx)
		elapsed := time.Since(start)
		fmt.Printf(fmt.Sprintf("Calculating burn took %s \n", elapsed))

		fmt.Println("Calculating months of runway for t= 1 to 24")
		start = time.Now()
		for i := 0; i < num; i++ {
			o := months_of_runway(incr.Root(), i)
			incr.Observe(incr.Root(), graph, o)
		}

		graph.Stabilize(ctx)
		elapsed = time.Since(start)
		fmt.Printf(fmt.Sprintf("Calculating months of runway took %s \n", elapsed))

		assert.Fail(t, "deliberate")
	})
Calculating burn for t= 1 to 24
Calculating burn took 798.458µs 
Calculating months of runway for t= 1 to 24
Calculating months of runway took 10.725125ms 
```
Here, calculating burn is super fast and in this order, calculating months of runway is faster than test 1 because I assume cached results stepped in.

**Comment by wcharczuk** [comment]
If you pull latest main, I've added this case as an "example" under `examples/integration`.

Specifically, if you look at the `main.go` file in there, I've changed the `burn` function slightly:

```go
	// burn(t) = f(t)
	burn := func(bs *incr.BindScope, t int) incr.Incr[*int] {
		key := fmt.Sprintf("burn-%d", t)
		if _, ok := cache[key]; ok {
			return incr.WithinBindScope(bs, cache[key])
		}
		o := incr.Bind(bs, fakeFormula, func(bs *incr.BindScope, formula string) incr.Incr[*int] {
			return f(bs, t)
		})
		o.Node().SetLabel(key)
		cache[key] = o
		return o
	}
```

The change specifically is to "cache" the result of the burn nodes, similarly to how we cache the result of the `f` nodes.

The resulting speeds are then roughly equivalent first,second vs. second,first.

Digging in deeper, there actually was a bug in how the "ShouldRecompute" state of bind nodes was propagated; it wasn't considering if you just re-bound the same node that the computation should stop (instead it continued, causing children to recompute even though the parent technically didn't change).

With the fixed version of ShouldRecompute, and the above change, I think you'll be in good shape.

Going to close this out.

**Comment by wcharczuk** [comment]
Ok so the above _is_ technically true, but there was another bug.

Because we link the bound nodes to the bind itself, and then that remains the only link to the nodes that take the bind, we can get into situations where the bound node recomputing will cause the bind to re-bind unnecessarily, as we only care about running the bind function if the bind's input changes. 

I've tweaked Bind to only consider the changedAt of the _input_ node when it stabilizes, and this should fix the above *without needing to use the cache trick*.

I've also reflected this in the integration example.
